File: lesson-570-mllib-graphx/solutions/python/solution-01-ml-kmeans/solution.py
from numpy import array
from math import pow

from pyspark.mllib.clustering import KMeans, KMeansModel

srcDir = "file:///Users/kmusial/hadoopupdate/data"

file = srcDir + "/tx.csv"
text = sc.textFile(file)

txns = text.map(lambda st: st.split(",")).map(lambda el: (el[0], el[1], float(el[2])))

# get the amounts for clustering training
amnts = txns.map(lambda v: array([v[2]]))

clusterCount = 10
iterationCount = 20
model = KMeans.train(amnts, clusterCount, iterationCount)

print("Cluster centers")
for c, value in enumerate(model.centers):
    print(c, value[0])

# generate the amounts for clustering: +- 5*10^i, i=0..3
vals = sc.parallelize(xrange(4)).map(lambda i: pow(10, i)).flatMap(lambda x: array([-5*x, 5*x])).map(lambda v: array([v]))

# predict the clusters for the amounts
clsd = vals.map(lambda s: (s, model.predict(s)))

print("Predictions")
for value, c in clsd.collect():
    print(value[0], c)

File: lesson-570-mllib-graphx/solutions/scala/solution-01-ml-kmeans/solution.scala

import org.apache.spark.ml._
import org.apache.spark.ml.clustering._
import org.apache.spark.ml.feature._
import scala.math._

case class Tx(date: String, desc: String, amount: Double)
val training = sc.textFile(srcDir + "/tx.csv").map(_.split(",")).map(x => Tx(x(0), x(1), x(2).toDouble)).toDF.cache

val assembler = new VectorAssembler().setInputCols(Array("amount")).setOutputCol("features")
val k = 4
val maxIterations = 10000
val km = new KMeans().setMaxIter(maxIterations).setK(k)
val pl = new Pipeline().setStages(Array(assembler, km))

val model = pl.fit(training)

val centers = model.stages(1).asInstanceOf[KMeansModel].clusterCenters
println("Cluster Centers:")
centers.zipWithIndex.map(c => (c._2,c._1(0))).foreach(println(_))

case class Val(amount:Double)
val seq = List.range(0,4).map(pow(10, _)).flatMap(x => Array(Val(-5*x),Val(5*x)))
val test = sqlContext.createDataFrame(seq)

val result = model.transform(test)
println("Predictions:")
result.show()
